contrib/versionfile/src/mill/contrib/versionfile/VersionFileModule.scala

```diff
@@ -0,0 +1,44 @@
+package mill.contrib.versionfile
+
+import mill._, scalalib._
+
+trait VersionFileModule extends Module {
+
+  implicit val wd = os.pwd
+
+  def versionFileName = "version"
+  def versionFileDirectory = wd
+  def versionFilePath = versionFileDirectory / versionFileName
+
+  def currentVersion = T.input { Version.of(os.read(versionFilePath)) }
```

Comment:
```suggestion
  /** The file containing the current version. */
  def versionFile: Source = T.source(millSourcePath / "version")
  /** The current version. */
  def currentVersion: T[Version] = T { Version.of(os.read(versionFilePath().path)) }
```

That way, we avoid reading the file if it hasn't changed. It looks also a bit more idiomatic (if that is even a thing).

Reply:
Won't `millSourcePath / "version"` prevent one from having the version file at root level of the repository?

Reply:
Yeah. It's an oversight from me. But `wd` isn't mill API. You can either override `millSourcePath` or you can use the value `millOuterCtx.millSourcePath`.